### PR TITLE
[Bugfix] fixed leak in SpMMCreateBlocks

### PR DESCRIPTION
**Pull request by Kacper-Pietkun**
## Description

Fixed memory leak that happened in SpMMCreateBlocks

memory usage before fix - constantly growing with each epoch
image

memory usage after fix - flat chart
image

Charts were generated while running graphsage model (examples/pytorch/ogb/ogbn-products/graphsage/main.py) for 14 epochs and with cpu

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).

### CODE REVIEW

**Comment by bartekkuncer** [comment]
image
Images seem to be broken.

**Comment by Kacper-Pietkun** [comment]
> image Images seem to be broken.

It seems fine for me.

**Comment by bartekkuncer** [comment]
> > image Images seem to be broken.
> 
> It seems fine for me.

For some reason I can only see them after clicking on the image link. I guess there is a problem on my side. Sorry for the false alarm.

**Comment by yaox12** [comment]
@dgl-bot

File: src/array/cpu/spmm_blocking_libxsmm.h [hunk 1]
```diff
@@ -80,10 +80,16 @@ inline void SpMMCreateBlocks(
...
+    IdType *indices_block_buf = nullptr;
+    if (use_lhs) {
+      indices_block_buf = reinterpret_cast<IdType *>(
+          aligned_alloc(64, indptr[M] * sizeof(IdType)));
+    }
+    IdType *edges_block_buf = nullptr;
+    if (use_rhs) {
+      edges_block_buf = reinterpret_cast<IdType *>(
+          aligned_alloc(64, indptr[M] * sizeof(IdType)));
+    }
```
Comment (BarclayII) [inline src/array/cpu/spmm_blocking_libxsmm.h:92]: Could you add a comment on why this change fixes the memory leak?  As far as I remember `aligned_alloc(3)` doesn't require using `free(3)` to reclaim memory (https://linux.die.net/man/3/aligned_alloc).  Since I didn't see any `free(3)` call I think there should be other reasons.

Comment (Kacper-Pietkun) [reply src/array/cpu/spmm_blocking_libxsmm.h:92]: Memory, which is  allocated here:
https://github.com/dmlc/dgl/pull/5210/files#diff-09f913927175aed9edd405df276b75d0c56b07ca917a887ab5ed19344e960f07R83-R92
image

gets deallocated here:
https://github.com/dmlc/dgl/pull/5210/files#diff-09f913927175aed9edd405df276b75d0c56b07ca917a887ab5ed19344e960f07R401-R411
image

As you can see memory is deallocated only when "use_lhs" or "use_rhs" flags are equal to "true". So to avoid a leak, appropriate "if statements" have to be added before allocation.

aligned_alloc requires using free in order to avoid a memory leak according to the spec https://en.cppreference.com/w/c/memory/aligned_alloc

**Review by bgawrych — APPROVED** [review]

**Review by BarclayII — APPROVED** [review]

**Comment by BarclayII** [comment]
@dgl-bot
